formulaic/transforms/basis_spline.py

```diff
@@ -122,10 +125,18 @@ def basis_spline(  # pylint: disable=dangerous-default-value  # always replaced
             "Some field values extend beyond upper and/or lower bounds, which can result in ill-conditioned bases. "
             "Pass a value for `extrapolation` to control how extrapolation should be performed."
         )
-    if extrapolation is SplineExtrapolation.CLIP:
-        x = numpy.clip(x, lower_bound, upper_bound)
-    if extrapolation is SplineExtrapolation.NA:
-        x = numpy.where((x >= lower_bound) & (x <= upper_bound), x, numpy.nan)
+    if extrapolation in (
+        SplineExtrapolation.CLIP,
+        SplineExtrapolation.NA,
+        SplineExtrapolation.ZERO,
+    ):
+        valid_x = (x >= lower_bound) & (x <= upper_bound)
+        if extrapolation is SplineExtrapolation.CLIP:
+            x = numpy.clip(x, lower_bound, upper_bound)
+        elif extrapolation is SplineExtrapolation.NA:
+            x = numpy.where((x >= lower_bound) & (x <= upper_bound), x, numpy.nan)
+    else:
+        valid_x = numpy.ones_like(x, dtype=bool)
```

Comment:
nit: It would likely be (in most cases only marginally) cheaper to pre-compute the data degrees of freedom here, rather than building a new ones array in memory.

Reply:
Actually, would it make sense to just remove this dof exception altogether? Sure, it may not make a great deal of sense for small datasets, but quantile does interpolation, and it probably stops making sense a lot earlier than when the row count is less than the number of knots... But it might be better to just output splines anyway, so that the same algorithm doesn't fail seemingly randomly on datasets with different cardinalities. Not sure if you have strong opinions about this.

Reply:
I have rewritten things so that if extrapolation is extend that there is no selection at all.  It added a bit more cyclomatic complexity, but I think it reduces temp storage and in many cases (when LB and UB are set to min() and max()) will completely eliminate any copying.